triedb/pathdb/journal.go

```diff
@@ -226,8 +235,27 @@ func (db *Database) loadDiffLayer(parent layer, r *rlp.Stream) (layer, error) {
 	for i, addr := range jaccounts.Addresses {
 		accounts[addr] = jaccounts.Accounts[i]
 	}
-	if err := r.Decode(&jstorages); err != nil {
-		return nil, fmt.Errorf("load diff storages: %v", err)
+	if layerJournalVersion == journalVersionV0 {
+		var jstoragesV0 []journalStorageV0
+		if err := r.Decode(&jstoragesV0); err != nil {
+			return nil, fmt.Errorf("load diff storages: %v", err)
+		}
+		jstorages = make([]journalStorage, 0, len(jstoragesV0))
+		for _, st := range jstoragesV0 {
+			if st.Incomplete { // Storage diff entries that are complete are compatible with the journal v1 type.
+				log.Warn("legacy v0 diff layer shows incomplete storage-diff write, cannot recover this, have to drop journal")
+				return nil, fmt.Errorf("legacy v0 diff layer with incomplete storage-diff: %w", errUnexpectedVersion)
+			}
+			jstorages = append(jstorages, journalStorage{
+				Account: st.Account,
+				Hashes:  st.Hashes,
+				Slots:   st.Slots,
+			})
+		}
+	} else {
+		if err := r.Decode(&jstorages); err != nil {
+			return nil, fmt.Errorf("load diff storages: %v", err)
+		}
```

Comment:
nit, could remove unnecessary indentation
```suggestion
	} else if err := r.Decode(&jstorages); err != nil {
		return nil, fmt.Errorf("load diff storages: %v", err)
```